frontend/src/components/utils.js

```diff
@@ -8,124 +8,154 @@ export const flatData = (data) => {
+ * Takes the payload, filters the denormalized data to make sure students do not
+ * appear in more than one section. Returns an array of rubric data based on assignment.
+ * @param {array} payload all the data that is provided to state.payload
+ * @returns array
+ */
+export const pivotHeatMapDataNoSections = (payload) => {
+  // filter denormalized data so students dont appear in more than one section
+  const filteredDenormalizedData = payload.denormalized_data.filter(uniqueStudentCriteria);
+
+  const allRubrics = payload.assignments.map((assignment) => {
+    const rubric = {
+      assignmentId: assignment.id,
+      name: assignment.name,
+      totalAssessments: 0,
+      dataPoints: assignment.rubric.map((criterion) => criterion.ratings.map((rating) => {
+        const dataPoint = {
+          criterionId: criterion.id,
+          criterion: criterion.description,
+          ratingDescription: rating.description,
+          maxPoints: rating.points,
+          count: filteredDenormalizedData.reduce((acc, curr) => (
+            acc + Number(curr.criterion_id === `${assignment.id}${criterion.id}`
+              && curr.rating === rating.description)), 0),
+        };
+        return dataPoint;
+      })),
+    };
+    rubric.totalAssessments = sumTotalAssessments(rubric);
+    rubric.dataPoints = calculateDataPointValues(rubric.dataPoints, rubric.totalAssessments);
+    return rubric;
+  });
+  return allRubrics;
+};
+
+/**
+ * Takes the payload, pivots the data to represent rubrics based on assigment and section.
+ * @param {array} payload all the data that is provided to state.payload
+ * @returns array
+ */
+export const pivotHeatMapData = (payload) => {
+  const sections = sectionIdNameMap(payload.sections);
 
   const allRubrics = payload.assignments.map((assignment) => {
     const rubrics = [];
-    sections.forEach((section) => {
+    sections.forEach((sectionObject) => {
       const rubric = {
         assignmentId: assignment.id,
         name: assignment.name,
         dueDate: assignment.due_at,
-        sectionId: section,
-        totalAssessments: payload.submissions.map((sub) => sub.submissions)
-          .flat()
-          .reduce((acc, curr) => (
-            acc + Number(assignment.id === curr.assignment_id
-              && studentLookup[curr.user_id] === section)
-          ), 0),
+        sectionId: sectionObject.sectionId,
+        sectionName: sectionObject.sectionName,
+        totalAssessments: 0,
         dataPoints: assignment.rubric.map((criterion) => criterion.ratings.map((rating) => {
           const dataPoint = {
             criterionId: criterion.id,
             criterion: criterion.description,
             ratingDescription: rating.description,
             maxPoints: rating.points,
             count: payload.denormalized_data.reduce((acc, curr) => (
-              acc + Number(curr.criterion_id === criterion.id
+              acc + Number(curr.criterion_id === `${assignment.id}${criterion.id}`
```

Comment:
Yeah, I think it would be a good idea. We'd have to loop over every piece of data that has criterion ids, and it might slow down the delivery of the payload, but its probably worth it.